tools/sof_ri_info/sof_ri_info.py

```diff
@@ -0,0 +1,883 @@
+        ext_idx += 1
+
+    return css_mft
+
+def parse_mft_extension(reader, ext_id):
+    """ Parses mft extension from sof binary
+    """
+    ext_type = reader.read_dw()
+    ext_len = reader.read_dw()
+    if ext_type == 15:
+        ext = PlatFwAuthExtension(ext_id, reader.get_offset()-8)
+        reader.ff_data(ext_len-8)
+    elif ext_type == 17:
+        ext = AdspMetadataFileExt(ext_id, reader.get_offset()-8)
+        ext.add_a(Auint('adsp_imr_type', reader.read_dw(), 'red'))
+        # skip reserved part
+        reader.read_bytes(16)
+        reader.read_dw()
+        reader.read_dw()
+        #
+        ext.add_a(Auint('version', reader.read_dw()))
+        ext.add_a(Abytes('sha_hash', reader.read_bytes(32)))
+        ext.add_a(Auint('base_offset', reader.read_dw()))
+        ext.add_a(Auint('limit_offset', reader.read_dw()))
+        ext.add_a(Abytes('attributes', reader.read_bytes(16)))
+    else:
+        ext = MftExtension(ext_id, 'Other Extension', reader.get_offset()-8)
+        reader.ff_data(ext_len-8)
+    ext.add_a(Auint('type', ext_type))
+    ext.add_a(Auint('length', ext_len))
+    return ext
+
+def parse_adsp_manifest_hdr(reader):
+    """ Parses ADSP manifest hader from sof binary
+    """
+    # Verify signature
+    try:
+        sig = reader.read_bytes(4).decode()
+    except UnicodeDecodeError:
+        print('\n' + reader.offset_to_string() + \
+              '\terror: Failed to decode signature, wrong position?')
+        sys.exit(1)
+    if sig != '$AM1':
+        reader.error('ADSP Manifest NOT found!', -4)
+        sys.exit(1)
+    reader.info('ADSP Manifest (' + sig + ')', -4)
+
+    hdr = Component('adsp_mft_hdr', 'ADSP Manifest Header',
+                    reader.get_offset() -4)
+    hdr.add_a(Astring('sig', sig))
+
+    hdr.add_a(Auint('size', reader.read_dw()))
+    hdr.add_a(Astring('name', chararr_to_string(reader.read_bytes(8), 8)))
+    hdr.add_a(Auint('preload', reader.read_dw()))
+    hdr.add_a(Auint('fw_image_flags', reader.read_dw()))
+    hdr.add_a(Auint('feature_mask', reader.read_dw()))
+    hdr.add_a(Aversion('build_version', reader.read_w(), reader.read_w(),
+                       reader.read_w(), reader.read_w()))
+
+    hdr.add_a(Adec('num_module_entries', reader.read_dw()))
+    hdr.add_a(Ahex('hw_buf_base_addr', reader.read_dw()))
+    hdr.add_a(Auint('hw_buf_length', reader.read_dw()))
+    hdr.add_a(Ahex('load_offset', reader.read_dw()))
+
+    return hdr
+
+def parse_adsp_manifest_mod_entry(index, reader):
+    """ Parses ADSP manifest module entry from sof binary
+    """
+    # Verify Mod Entry signature
+    try:
+        sig = reader.read_bytes(4).decode()
+    except UnicodeDecodeError:
+        print(reader.offset_to_string() + \
+              '\terror: Failed to decode ModuleEntry signature')
+        sys.exit(1)
+    if sig != '$AME':
+        reader.error('ModuleEntry signature NOT found!')
+        sys.exit(1)
```

Comment:
Wouldn't exception be better?

Reply:
Standalone, single purpose tool... I'd say it's debatable.

Reply:
Still easier to find error, when you get stacktrace instead of just message and exiting program

Reply:
In this case, the context of error is so obvious that I assumed simple error message to more readable than the stack trace. This is the reason of catching decode exceptions in other places as well.